hpx/util/unwrap.hpp

```diff
@@ -0,0 +1,166 @@
+//  Copyright (c) 2017 Denis Blank
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef HPX_UTIL_UNWRAP_HPP
+#define HPX_UTIL_UNWRAP_HPP
+
+#include <hpx/config.hpp>
+#include <hpx/util/detail/unwrap_impl.hpp>
+
+#include <cstddef>
+#include <utility>
+
+namespace hpx {
+namespace util {
+    /// A helper function for retrieving the actual result of
+    /// any hpx::lcos::future like type which is wrapped in an arbitrary way.
+    ///
+    /// Unwraps the given pack of arguments, so that any hpx::lcos::future
+    /// object is replaced by its future result type in the argument pack:
+    /// - `hpx::future<int>` -> `int`
+    /// - `hpx::future<std::vector<float>>` -> `std::vector<float>`
+    /// - `std::vector<future<float>>` -> `std::vector<float>`
+    ///
```

Comment:
Sadly the underlying API is only capable of doing a synchronous unwrap which isn't suitable for the current implementation of `when_all`. 
In order to improve `when_all` we would have to use a stackful coroutine with this API or
implement another version of the traversal API which is completely stateless (to emulate a stackless coroutine).